### ISSUE TITLE
noticed strange behavior, not sure if it using coroutineScope incorrectly

### ISSUE THREAD

**Issue by lannyf77** [Kotlin/kotlinx.coroutines#1359]
Having a list of the data processors.  Each data processor has its own `dpProcess()` which returns true if the process successfully done.

all of the `processor.dpProcess()`  run in parallel with others (via the launch builder).

but only after all data porcessors are completed then the final  `onAllDataProcessed()`  should be called.
This is achived by the coroutineScope {} block.

and also the coroutineScope {} block is controled by the `withTimeoutOrNull(2000)`, so that if the processing of data is not completed in 2 sec then will abort it

        suspend fun processData(): Boolean {
            var allDone = dataProcesserList.size > 0
            val result = withTimeoutOrNull(2000) {  // should timeout at 2000
                coroutineScope {  // block the execution sequence until all sub coroutines are completes
                    for (processor in dataProcesserList) {
                        launch { // launch each individual process in parallel
                            allDone = allDone && processor.dpProcess()  // update the global final allDone,   if anyone returns false it should be return false 
                        }
                    }
                }
            }

            /**
             * strange:  one of the processor.dpProcess() has return false,  but the allDone at here still shows true (and it is not timed out)
             */

            if (result != null && allDone) { // not timeout and all process completed successfully
                onAllDataProcessed()
            } else {
                allDone = false  // timeout or some process failed
            }
            return allDone
        }

Noticded a strange behavior, and happens more on device of Android 7.1.1 device than others

Some time the final allDone is true even one of the `processor.dpProcess()`  returns false and not timed out.
But same case sometime returns false as expected

Is it because that how it is coded which brings this behavior?

What is better way to put a few functions to run in parallel and wait until all of them are done then advanced to next step. 
And the whole process should timeout and abort

**Comment by RBusarow** [comment]
The problem is that `allDone` is a shared mutable state.  You have all you coroutines updating the same Boolean simultaneously, but that involves a read and a write.

Assume that you have two coroutines, **a** and **b**.  **a** wants to report a failure but **b** succeeds.

coroutine **a** --> read `allDone` (true)
coroutine **b** --> read `allDone` (true)
coroutine **a** --> write `allDone` (true && false -> false)
coroutine **b** --> write `allDone` (true && true -> true)

Any further reads of `allDone` will be unaware that its value was ever set to false.

We should also look at the scheduling of when you're doing your reads and writes.
```Kotlin
allDone = allDone && processor.dpProcess()
```
The initial read of `allDone` happens eagerly.  But the write of the new value to `allDone` is suspended until after `dpProcess()` is complete.  If `dpProcess()` takes a while to execute, then every coroutine created in the for loop will have already read that same initial value of `true`.

You can find a variety of solutions to handling shared mutable state here: https://kotlinlang.org/docs/reference/coroutines/shared-mutable-state-and-concurrency.html

**Comment by lannyf77** [comment]
@RBusarow  thanks for point it out.  

seems this works,  but not sure if it could be simplified better (dont like the joinAll() then using aWait() for the result) 
 and what is best way to handle exception here, using  try/catch ?  but if timeout the aWait() will not throw, right?

       suspend fun processData(): Boolean = withContext(Dispatchers.IO) {
    
        var allDone = dataProcesserList.size > 0
        if (!allDone) {
            allDone
        } else {
    
            val jobs: List<Deferred<Boolean>> = dataProcesserList.map {processor->
                async(start = CoroutineStart.LAZY) {
                    processor.dpProcess()
                }
            }
            val result = withTimeoutOrNull(2000) {
                jobs.joinAll()
            }
            if (result != null) {
                allDone = jobs.all { deferred ->
                    deferred.await()
                }
                if (allDone) onAllDataProcessed()
            } else {
                allDone = false
            }
            allDone
        }
    }

**Comment by fvasco** [comment]
@lannyf77 you can consider something simpler:

```kotlin
suspend fun processData(): Boolean {
    val allDone = withTimeoutOrNull(2_000) {
        dataProcesserList.map {
            async { it.dpProcess() }
        }
                .awaitAll()
                .all { it }
    }
            ?: false

    if (allDone) onAllDataProcessed()

    return allDone
}
```

**Comment by lannyf77** [comment]
@fvasco, thanks for the simpler one!   what about the consideration of handling exception?
and saw https://stackoverflow.com/questions/55085363/launch-a-number-of-coroutines-and-join-them-all-with-timeout-without-cancelling for similar case, using `withTimeoutOrNull(timeoutMillis) { jobs.joinAll() }` for the reason of  ```We use joinAll (as opposed to awaitAll) to avoid exception if one of the jobs fail and withTimeoutOrNull to avoid exception on timeout.```

**Comment by fvasco** [comment]
Hi @lannyf77,
it works **as designed**.
I suggest you to find the answer yourself, get it on https://kotlinlang.org/docs/reference/coroutines/composing-suspending-functions.html#structured-concurrency-with-async

Moreover I suggest you to take a look at #320

Make some test in a playground an come back when you need a more specific question.
